src/main/scala/decider/Z3ProverAPI.scala

```diff
@@ -0,0 +1,439 @@
+    "nlsat.seed" -> 0,
+    "random_seed" -> 0,
+  )
+  val stringParams = Map(
+    //"smt.qi.cost" -> "(+ weight generation)", // cannot set this for some reason, but this is the default value anyway.
+    "sat.phase" -> "caching"
+  )
+  val doubleParams = Map(
+    "smt.restart_factor" -> 1.5,
+    "smt.qi.eager_threshold" -> 100.0,
+  )
+}
+
+
+class Z3ProverAPI(uniqueId: String,
+                  termConverter: TermToZ3APIConverter,
+                  identifierFactory: IdentifierFactory,
+                  reporter: Reporter)
+    extends Prover
+      with LazyLogging
+{
+
+  protected var pushPopScopeDepth = 0
+  protected var lastTimeout: Int = -1
+  protected var logfileWriter: PrintWriter = _
+  protected var prover: Solver = _
+  protected var ctx: Context = _
+
+  var proverPath: Path = _
+  var lastModel : Model = _
+
+  var emittedPreambleString = mutable.Queue[String]()
+  var preamblePhaseOver = false
+  var preambleAssumes = mutable.LinkedHashSet[BoolExpr]()
+  val emittedSorts = mutable.LinkedHashSet[com.microsoft.z3.Sort]()
+  val emittedSortSymbols = mutable.LinkedHashSet[Symbol]()
+  val emittedFuncs = mutable.LinkedHashSet[FuncDecl]()
+  val emittedFuncSymbols = mutable.Queue[Symbol]()
+
+  val expandMacros = true
+
+
+
+  def version(): Version = {
+    var versString = com.microsoft.z3.Version.getFullVersion
+    if (versString.startsWith("Z3 "))
+      versString = versString.substring(3)
+    Version(versString)
+  }
+
+  def start(): Unit = {
+    pushPopScopeDepth = 0
+    lastTimeout = -1
+    logfileWriter = if (Verifier.config.disableTempDirectory()) null else viper.silver.utility.Common.PrintWriter(Verifier.config.proverLogFile(uniqueId).toFile)
+    ctx = new Context(Z3ProverAPI.initialOptions.asJava)
+    val params = ctx.mkParams()
+    Z3ProverAPI.boolParams.foreach{
+      case (k, v) => params.add(k, v)
+    }
+    Z3ProverAPI.intParams.foreach{
+      case (k, v) => params.add(k, v)
+    }
+    Z3ProverAPI.doubleParams.foreach{
+      case (k, v) => params.add(k, v)
+    }
+    Z3ProverAPI.stringParams.foreach{
+      case (k, v) => params.add(k, v)
+    }
+    prover = ctx.mkSolver()
+    prover.setParameters(params)
+    termConverter.start()
+    termConverter.ctx = ctx
+    emittedPreambleString.clear()
+    preamblePhaseOver = false
+    emittedFuncs.clear()
+    emittedSorts.clear()
+    emittedFuncSymbols.clear()
+    emittedSortSymbols.clear()
+  }
+
+  def reset(): Unit = {
+    stop()
+    termConverter.reset()
+    start()
+  }
+
+  def stop(): Unit = {
+    if (ctx != null){
+      ctx.close()
+      ctx = null
+      emittedPreambleString.clear()
+      preamblePhaseOver = false
+      emittedFuncs.clear()
+      emittedSorts.clear()
+      emittedFuncSymbols.clear()
+      emittedSortSymbols.clear()
+    }
+  }
+
+  def push(n: Int = 1): Unit = {
+    endPreamblePhase()
+    pushPopScopeDepth += n
+    if (n != 1)
+      sys.error("Not supported")
+    prover.push()
```

Comment:
I assume `n > 1` is simply not used, although I guess we could have just had a loop here. Still, could the `n` arguments simply be removed everywhere?

Reply:
I haven't seen it used anywhere, but there might be a special case that does (Malte would know I guess). I'll just add a loop though, the not supported case is a leftover from when I first tried this out and just wanted to get it running, and I just forgot to fix it.